architecture-docs/adr-003.md

```diff
@@ -0,0 +1,54 @@
+# ADR 003: Migrate to jellyfish merkle trie
+
+## Changelog
+* 31-03-2020: Initial Draft
+
+## Context
+
+Need a merkle trie library to store staking state, current one has several disadvantages:
+
+- Written and maintained by single person.
+- Custom data structure and algorithm.
+- Not enough documentation.
+
+## Decision
+
+### jellyfish merkle trie
+
+https://github.com/libra/libra/blob/master/storage/jellyfish-merkle/src/lib.rs
+
+#### Advantages
+
+- Review from Facebook security engineers
+- Small code base, well documented.
+- Well known data structure (simplified version of ethereum MPT).
+- Easy to integrate.
+- Optimized for pbft: Use u64 version number instead of hash.
+- License is apache 2.0.
+- Support inclusion proof and range proof.
+
+#### Disadvantages
+
+- Tied with a lot of custom Libra crates, need to maintain a separated library.
+- No extension node compared to ethereum MPT, deeper tree in theory, a trade off made in favor of simplicity.
+
+### Integration
+
+- Maintain a custom repository with code separated from libra repo.
```

Comment:
more details needed -- e.g. are all the other crates like https://github.com/yihuang/libra-jellyfish-merkle/tree/master/crypto and all its transitive dependencies needed?
If this is a custom version updated with a script (instead of git merging), why not to include it in the main repo + have the rsync script?

Reply:
One of the concern would be whether it is manageable to keep in sync with the master from time-to-time. I have a quick look on the commit history, it seems it is relatively stable since Dec.

Do you know what is their plan? Is the implementation freezed?

Reply:
not sure if there's a public code roadmap, but seems fairly stable
As for signatures -- they shouldn't be needed for the jellyfish-merkle anyway?

Reply:
Oh sorry, for signature I am referring to the method signature. But yes  it seems to be relatively stable.

Reply:
From what I see, there is actually no clear winner that is good enough for us to use as a black box. Even if it's written by Facebook, it's still pre-alpha software in theory. And the code syncing and patching process is also error prone.
I think no matter we choose jellyfish or starling (or other solution), we need to write down the algorithm as a specification, to make sure we all understand the implications.

Reply:
I am not sure if this will be overkilled. But can we implement an extra abstraction layer between our code and whatever lib we decided to use? That way we are free from changes from the underlying library.

When there are breaking changes we will still be affected, but we lock down the potion of code we will have to change.

Reply:
> I am not sure if this will be overkilled. But can we implement an extra abstraction layer between our code and whatever lib we decided to use? That way we are less dependent with changes from the underlying library.
> 
> When there are breaking changes we will still be affected, but at least we lock down the portion of code we will have to change.

That's already kind of done -- "persistent storage" is abstracted out in `kvdb`; "updates and lookups from merkle trie" are in buffer. 
but it should be documented in ADR how exactly this other crate is planned to be integrated in -- https://github.com/crypto-com/chain/pull/1330#discussion_r400715258

Reply:
> more details needed -- e.g. are all the other crates like https://github.com/yihuang/libra-jellyfish-merkle/tree/master/crypto and all its transitive dependencies needed?
> If this is a custom version updated with a script (instead of git merging), why not to include it in the main repo + have the rsync script?

We can remove these dependencies like crypto/crypto-derive/lcs by extracting needed code into the merkle-trie crate, the result is easier to maintain by us, but harder to sync code with upstream automatically.

Reply:
> the result is easier to maintain by us, but harder to sync code with upstream automatically.

that's fine -- but should be mentioned in the impact; even with a script, it may become hard with some external repo restructuring (as opposed to maintaining a fork and using git merges/patches)

Reply:
Updated

- Use a fork of the whole libra repository, more convenient to sync the code in the future with git.
- Describe more detailed integration work.

Reply:
The integration code:
https://github.com/yihuang/chain/blob/issue802/chain-storage/src/jellyfish.rs